File: paleo_workbench/mapping/crs_contract.py
"""Unified CRS contract — the single CRS predicate/resolution authority (V9 W3).

Before this module the workbench carried four coexisting CRS predicates /
normalizers (``workflow.crs_policy``, ``map_render_backend._normalize_crs_name``,
``reference_layers._normalize_crs_input``, ``qgis_mirror._geographic_auth``)
plus a handful of "quietly EPSG:4326" defaults. This module does NOT
re-implement any axis math — it is the one place that answers:

* **declared?** — ``normalize_crs`` delegates to the render backend's single
  normalizer (auth-id canonicalization);
* **geographic?** — ``crs_is_geographic`` delegates to
  :func:`paleo_workbench.workflow.crs_policy.crs_is_geographic`, the single
  axis-units truth (known-set + pyproj; ``None`` = unverifiable);
* **what happens when a CRS is missing** — :func:`resolve_crs` returns a
  :class:`CRSResolution` that either carries the declared CRS or an explicit
  degraded reason. A silent ``EPSG:4326`` assumption is a contract violation:
  undeclared stays undeclared (``crs=""``) and the degraded state is recorded
  by the caller (mirror diagnostics, tool-context facts, measurement labels).

Leaf module: standard library + optional pyproj only.
"""
from __future__ import annotations

from functools import lru_cache

import logging
from dataclasses import dataclass

__all__ = [
    "CRSInference",
    "CRSResolution",
    "DomainMismatch",
    "coordinate_domain_mismatch",
    "crs_axis_unit_metres",
    "crs_coordinate_domain",
    "crs_is_geographic",
    "geod_for_crs",
    "infer_crs_from_extent",
    "normalize_crs",
    "panel_publish_crs",
    "resolve_crs",
    "scale_denominator_from_pixels",
]

logger = logging.getLogger(__name__)


def normalize_crs(crs: object) -> str:
    """Canonical CRS auth id (``""`` when absent/unparseable).

    Single normalization implementation: delegates to the render backend's
    ``_normalize_crs_name`` (pyproj-backed, historically the most complete
    normalizer in the tree). Every CRS comparison in the mapping stack goes
    through this — raw string compare against "EPSG:4326" style literals is
    the bug class this contract exists to remove.
    """
    text = str(crs or "").strip()
    if not text:
        return ""
    from paleo_workbench.mapping.map_render_backend import _normalize_crs_name

    return _normalize_crs_name(text)


def crs_is_geographic(crs: object) -> bool | None:
    """Axis-units truth (delegate): ``None`` = unverifiable, never a guess."""
    from paleo_workbench.workflow.crs_policy import crs_is_geographic as _axis_truth

    return _axis_truth(str(crs or "").strip() or None)


@dataclass(frozen=True, slots=True)
class CRSResolution:
    """The outcome of resolving a CRS requirement at one consumption site.

    ``crs`` is the resolved auth id (``""`` = none resolved). ``declared``
    is False exactly when the source carried no usable CRS; the
    ``degraded_reason`` then states *what was done instead* — the caller is
    expected to surface it (diagnostics entry, status label, QA annotation),
    never to swallow it.
    """

    crs: str
    declared: bool
    degraded_reason: str = ""

    @property
    def ok(self) -> bool:
        return self.declared and bool(self.crs)


def resolve_crs(
    value: object,
    *,
    purpose: str,
    fallback: object = "",
) -> CRSResolution:
    """Resolve a CRS requirement without silent defaults (V9 contract).

    ``value`` non-empty → declared resolution (normalized). Empty → the
    ``fallback`` **only when it is a real, recorded legacy default** and the
    result is marked degraded with the purpose named in the reason; no
    fallback → undeclared resolution (``crs=""``). Either way the caller can
    explain what happened — the forbidden pattern is
    ``project_crs or "EPSG:4326"`` with no trace.
    """
    normalized = normalize_crs(value)
    if normalized:
        return CRSResolution(crs=normalized, declared=True)
    fallback_text = normalize_crs(fallback)
    if fallback_text:
        return CRSResolution(
            crs=fallback_text,
            declared=False,
            degraded_reason=(
                f"{purpose}: CRS 未声明，按记录在案的默认 {fallback_text} 处理（降级）"
            ),
        )
    return CRSResolution(
        crs="",
        declared=False,
        degraded_reason=f"{purpose}: CRS 未声明且无默认——按未知坐标处理（降级）",
    )


def _axis_unit_metres_uncached(text: str) -> bool | None:
    try:
        from pyproj import CRS

        parsed = CRS.from_user_input(text)
        axis = parsed.axis_info[0] if parsed.axis_info else None
        if axis is None:
            return None
        return str(axis.unit_name or "").strip().lower() in {
            "metre", "meter", "m",
        }
    except Exception:
        return None


@lru_cache(maxsize=64)
def _axis_unit_metres_cached(text: str) -> bool | None:
    return _axis_unit_metres_uncached(text)


def crs_axis_unit_metres(crs: object) -> bool | None:
    """True when the CRS horizontal axis unit is exactly metres.

    ``None`` = unverifiable (unknown id / no pyproj). Used by the honest
    scale-denominator derivation: a denominator computed from pixel metrics
    is only meaningful for metre axes, everything else reports unknown
    rather than a plausible-looking wrong number.

    V10 R5：按归一化字符串 lru_cache（pyproj 解析在 extent 高频链上，
    同一 CRS 反复解析是纯浪费；未知串也缓存——失败路径同样昂贵）。
    """
    text = str(crs or "").strip()
    if not text:
        return None
    return _axis_unit_metres_cached(text)


def panel_publish_crs(value: object, *, purpose: str = "图层面板发布") -> str:
    """面板/发布路径的 CRS：未声明返回 ""（按原坐标呈现）。

    V9 W3（review-1 P1-1 存量清理）：面板发布快照的 ``or "EPSG:4326"``
    在未声明时伪造坐标系。诚实语义 = 未声明就不设 CRS（渲染按原坐标、
    降级可查），工程模型的默认值照常声明 4326——本函数只消灭伪造路径。
    """
    resolution = resolve_crs(value, purpose=purpose)
    if not resolution.declared:
        logger.warning("%s（按原坐标呈现）", resolution.degraded_reason)
        return ""
    return resolution.crs


def scale_denominator_from_pixels(
    map_units_per_pixel: float,
    pixels_per_inch: float,
    crs: object,
) -> float:
    """Honest scale denominator from pixel geometry (metre axes only).

    Same formula QGIS uses for pixel canvases (``mupp × inches-per-metre ×
    dpi``). Returns ``0.0`` — unknown — whenever the CRS axis unit is not
    *verifiably* metres (degrees, US survey feet, unverifiable ids): a
    denominator derived under any other unit silently misstates the scale,
    which is exactly the failure mode this contract exists to prevent. The
    native path never uses this helper (``QgsMapCanvas::scale()`` is the
    authority there); it serves the fallback canvas only.
    """
    if crs_axis_unit_metres(crs) is not True:
        return 0.0
    if map_units_per_pixel <= 0.0 or pixels_per_inch <= 0.0:
        return 0.0
    return map_units_per_pixel * 39.370078740157481 * pixels_per_inch


def geod_for_crs(crs: object):
    """``pyproj.Geod`` for a geographic CRS, else ``None``.

    Consumers (measurement fallback, W8) use this to replace planar
    raw-map-unit distances with geodesic ones when — and only when — the
    axis truth says geographic. A ``None`` return keeps the honest planar
    label instead of guessing an ellipsoid.
    """
    if crs_is_geographic(crs) is not True:
        return None
    try:
        from pyproj import CRS, Geod

        ellipsoid = CRS.from_user_input(str(crs)).ellipsoid
        return Geod(
            a=ellipsoid.semi_major_metre, b=ellipsoid.semi_minor_metre)
    except Exception as exc:
        try:
            # WGS84 remains the geodesic default for unknown ellipsoids —
            # an explicit, reviewed fallback (V9 W8, review-1 P2-2)：降级
            # 记入日志而非静默替换。
            from pyproj import Geod

            logger.warning(
                "geod_for_crs: %r 椭球不可解析（%s）——测地计算回退 WGS84",
                crs, exc)
            return Geod("WGS84")
        except Exception:
            return None


# ---------------------------------------------------------------------------
# 坐标域契约（拓扑编辑迁移 M0 §6：进前域校验 + 推断锁定）
# ---------------------------------------------------------------------------

# 地理 CRS 的轴域是解析定义（经纬度全域），不是经验值。
_GEOGRAPHIC_DEGREE_DOMAIN = (-180.0, -90.0, 180.0, 90.0)
# 变换/面积比较的容差：投影域边界按面积适用范围给出，贴边数据不算失配。
_DOMAIN_EPSILON = 1e-9
# 投影 CRS 的 area_of_use 角点变换是近似框——域外扩比例（贴边真实数据
# 不误拦；明显域外数据仍失配）。
_PROJECTED_DOMAIN_SLACK = 0.02


def crs_coordinate_domain(crs: object) -> tuple[float, float, float, float] | None:
    """声明 CRS 的有效坐标域（CRS 自身坐标单位）；不可推导 → None。

    - 地理 CRS → 经纬度全域（轴定义，精确）。
    - 投影 CRS → pyproj ``area_of_use``（适用范围的经纬度框）变换回该 CRS
      的坐标单位；无 area_of_use / 无 pyproj / 变换失败 → None（诚实：
      无法验证 ≠ 通过）。
    - 未声明 → None（没有可校验的域）。
    """
    text = normalize_crs(crs)
    if not text:
        return None
    if crs_is_geographic(text) is True:
        return _GEOGRAPHIC_DEGREE_DOMAIN
    try:
        from pyproj import CRS, Transformer

        parsed = CRS.from_user_input(text)
        usage = parsed.area_of_use
        if usage is None or usage.west is None or usage.south is None \
                or usage.east is None or usage.north is None:
            return None
        forward = Transformer.from_crs(
            CRS.from_epsg(4326), parsed, always_xy=True)
        corners = [
            (usage.west, usage.south), (usage.east, usage.south),
            (usage.east, usage.north), (usage.west, usage.north),
        ]
        xs: list[float] = []
        ys: list[float] = []
        for lon, lat in corners:
            x, y = forward.transform(lon, lat)
            # 采样点落在变换定义域外（反常的 area_of_use）→ 无法给出
            # 可信域，而不是拿 inf 当边界。
            if not (abs(x) < 1e12 and abs(y) < 1e12):
                return None
            xs.append(x)
            ys.append(y)
        # 投影域是 area_of_use 角点变换的近似框：外扩 2% 容差，贴着适用
        # 范围边缘的真实数据（跨带边界/赤道边）不算失配；地理域是轴
        # 定义，不外扩。
        minx, maxx = min(xs), max(xs)
        miny, maxy = min(ys), max(ys)
        slack_x = (maxx - minx) * _PROJECTED_DOMAIN_SLACK
        slack_y = (maxy - miny) * _PROJECTED_DOMAIN_SLACK
        return (minx - slack_x, miny - slack_y, maxx + slack_x, maxy + slack_y)
    except Exception:
        return None


@dataclass(frozen=True, slots=True)
class DomainMismatch:
    """一次「声明 CRS 坐标域 vs 数据实际坐标范围」失配的事实。"""

    crs: str
    extent: tuple[float, float, float, float]
    domain: tuple[float, float, float, float]

    def describe(self) -> str:
        return (
            f"声明 CRS {self.crs} 的有效坐标域为 "
            f"x[{self.domain[0]:g}, {self.domain[2]:g}] / "
            f"y[{self.domain[1]:g}, {self.domain[3]:g}]，"
            f"数据实际坐标范围为 x[{self.extent[0]:g}, {self.extent[2]:g}] / "
            f"y[{self.extent[1]:g}, {self.extent[3]:g}]"
        )


def coordinate_domain_mismatch(
    crs: object,
    extent: object,
) -> DomainMismatch | None:
    """数据范围未完全落在声明 CRS 的有效坐标域内 → 失配事实；否则 None。

    无法推导坐标域（未声明/投影无 area_of_use/pyproj 缺席）→ None：
    域校验只拦截**可证明**的失配（fail-open 与镜像层 extent 检查同约定）。
    """
    if extent is None:
        return None
    try:
        xmin, ymin, xmax, ymax = (float(v) for v in tuple(extent)[:4])
    except (TypeError, ValueError):
        return None
    if not (xmin <= xmax and ymin <= ymax):
        return None
    domain = crs_coordinate_domain(crs)
    if domain is None:
        return None
    dminx, dminy, dmaxx, dmaxy = domain
    eps = _DOMAIN_EPSILON * max(1.0, abs(dmaxx - dminx), abs(dmaxy - dminy))
    fits = (
        xmin >= dminx - eps and xmax <= dmaxx + eps
        and ymin >= dminy - eps and ymax <= dmaxy + eps
    )
    if fits:
        return None
    return DomainMismatch(
        crs=normalize_crs(crs) or str(crs or ""),
        extent=(xmin, ymin, xmax, ymax),
        domain=domain,
    )


@dataclass(frozen=True, slots=True)
class CRSInference:
    """按数据坐标范围对工程 CRS 声明的推断结论（§6 推断锁定）。

    ``suggested_crs`` 为空 = 保持本地（不声明）；非空 = 可识别的 CRS，
    由用户确认后写入声明并锁定。``basis`` 是推断依据（诊断/对话文案用）。
    """

    suggested_crs: str
    basis: str

    @property
    def suggests_declaration(self) -> bool:
        return bool(self.suggested_crs)


def infer_crs_from_extent(extent: object) -> CRSInference:
    """首次导入数据的坐标范围推断（§6）。

    - 完全落在经纬度域内 → 数据本身是地理坐标，建议声明 EPSG:4326。
    - 超出经纬度域 → 本地/投影坐标，从范围无法识别具体投影 → 保持本地
      （未声明）；声明留待用户手动指定。
    - 范围不可用（空层/无效）→ 不推断，保持本地。
    """
    if extent is None:
        return CRSInference("", "数据坐标范围不可用——不推断")
    try:
        bounds = tuple(float(v) for v in tuple(extent)[:4])
        if not (bounds[0] <= bounds[2] and bounds[1] <= bounds[3]):
            return CRSInference("", "数据坐标范围不可用——不推断")
    except (TypeError, ValueError, IndexError):
        return CRSInference("", "数据坐标范围不可用——不推断")
    mismatch = coordinate_domain_mismatch("EPSG:4326", extent)
    if mismatch is None:
        return CRSInference(
            "EPSG:4326", "数据坐标完全落在经纬度域内（地理坐标）")
    return CRSInference("", "数据坐标超出经纬度域（本地/投影坐标）——保持本地")

File: libs/mapping_kernel/src/crs_contract.cpp
#include <pwb/mapping/crs_contract.hpp>

#include <algorithm>
#include <cctype>
#include <cmath>
#include <cstddef>
#include <cstdio>
#include <string>

namespace pwb::mapping {
namespace {

std::string trim(const std::string& s) {
    std::size_t a = 0;
    while (a < s.size() &&
           std::isspace(static_cast<unsigned char>(s[a]))) {
        ++a;
    }
    std::size_t b = s.size();
    while (b > a &&
           std::isspace(static_cast<unsigned char>(s[b - 1]))) {
        --b;
    }
    return s.substr(a, b - a);
}

// Python `\b` after the digit run: the next character must not be a word
// character. ASCII [A-Za-z0-9_] is the practical class for CRS strings;
// bytes >= 0x80 are conservatively treated as word characters (Python's
// unicode \w includes non-ASCII letters/digits), so an exotic suffix
// leaves the text unchanged rather than splitting it.
bool is_word_char(char c) {
    const unsigned char u = static_cast<unsigned char>(c);
    return std::isalnum(u) != 0 || c == '_' || u >= 0x80;
}

// Python f"{x:g}" and C "%g" share one formatting origin (6 significant
// digits, trailing zeros stripped, exponent thresholds equal).
std::string format_g(double v) {
    char buf[64];
    std::snprintf(buf, sizeof buf, "%g", v);
    return buf;
}

}  // namespace

std::string normalize_crs(const std::optional<std::string>& crs) {
    const std::string text = trim(crs.value_or(""));
    if (text.empty()) {
        return "";
    }
    // re.match(r"^(EPSG:\d+)\b", text, re.IGNORECASE)
    static constexpr char kPrefix[5] = {'E', 'P', 'S', 'G', ':'};
    bool head_ok = text.size() >= 5;
    for (std::size_t i = 0; head_ok && i < 5; ++i) {
        const char c = text[i];
        head_ok = c == kPrefix[i] ||
                  (c >= 'a' && c <= 'z' &&
                   static_cast<char>(c - 'a') + 'A' == kPrefix[i]);
    }
    if (head_ok) {
        std::size_t digits_end = 5;
        while (digits_end < text.size() && text[digits_end] >= '0' &&
               text[digits_end] <= '9') {
            ++digits_end;
        }
        // `\d+\b`: every shorter digit run would end right before a digit
        // (a word char), so only the full run can satisfy the boundary.
        if (digits_end > 5 &&
            (digits_end == text.size() || !is_word_char(text[digits_end]))) {
            return "EPSG:" + text.substr(5, digits_end - 5);
        }
    }
    return text;
}

CRSResolution resolve_crs(const std::optional<std::string>& value,
                          const std::string& purpose,
                          const std::optional<std::string>& fallback) {
    const std::string normalized = normalize_crs(value);
    if (!normalized.empty()) {
        CRSResolution out;
        out.crs = normalized;
        out.declared = true;
        return out;
    }
    const std::string fallback_text = normalize_crs(fallback);
    if (!fallback_text.empty()) {
        CRSResolution out;
        out.crs = fallback_text;
        out.declared = false;
        out.degraded_reason =
            purpose + ": CRS 未声明，按记录在案的默认 " + fallback_text +
            " 处理（降级）";
        return out;
    }
    CRSResolution out;
    out.declared = false;
    out.degraded_reason =
        purpose + ": CRS 未声明且无默认——按未知坐标处理（降级）";
    return out;
}

std::string panel_publish_crs(const std::optional<std::string>& value,
                              const std::string& purpose) {
    const CRSResolution resolution = resolve_crs(value, purpose);
    if (!resolution.declared) {
        // Python logs the degraded reason here; the kernel returns the
        // honest "" and leaves surfacing to the caller.
        return "";
    }
    return resolution.crs;
}

std::optional<bool> crs_axis_unit_metres(const std::optional<std::string>& crs) {
    const std::string text = trim(crs.value_or(""));
    if (text.empty()) {
        return std::nullopt;
    }
    // Python parses the CRS with pyproj and inspects the first horizontal
    // axis unit (metre/meter/m). The kernel has no pyproj — this is the
    // ImportError path: unverifiable, never guessed.
    return std::nullopt;
}

double scale_denominator_from_pixels(double map_units_per_pixel,
                                     double pixels_per_inch,
                                     const std::optional<std::string>& crs) {
    const std::optional<bool> axis_metres = crs_axis_unit_metres(crs);
    if (!axis_metres.has_value() || !*axis_metres) {
        return 0.0;
    }
    if (map_units_per_pixel <= 0.0 || pixels_per_inch <= 0.0) {
        return 0.0;
    }
    // Unreachable in the kernel today (no pyproj → axes never verifiably
    // metres) but kept line-faithful: any future axis authority lands
    // directly on the QGIS pixel-canvas formula.
    return map_units_per_pixel * 39.370078740157481 * pixels_per_inch;
}

std::optional<Domain> crs_coordinate_domain(const std::optional<std::string>& crs) {
    const std::string text = normalize_crs(crs);
    if (text.empty()) {
        return std::nullopt;
    }
    if (crs_is_geographic(text) == true) {
        // Axis-analytic domain: exact, no slack.
        return kGeographicDegreeDomain;
    }
    // Python transforms the projected CRS's area_of_use corners back into
    // CRS units (+2% box slack). Kernel: no pyproj → cannot derive →
    // nullopt (cannot-verify ≠ pass).
    return std::nullopt;
}

std::string DomainMismatch::describe() const {
    return "声明 CRS " + crs + " 的有效坐标域为 x[" + format_g(domain[0]) +
           ", " + format_g(domain[2]) + "] / y[" + format_g(domain[1]) +
           ", " + format_g(domain[3]) + "]，数据实际坐标范围为 x[" +
           format_g(extent[0]) + ", " + format_g(extent[2]) + "] / y[" +
           format_g(extent[1]) + ", " + format_g(extent[3]) + "]";
}

std::optional<DomainMismatch> coordinate_domain_mismatch(
    const std::optional<std::string>& crs, const std::optional<Domain>& extent) {
    if (!extent.has_value()) {
        return std::nullopt;
    }
    const double xmin = (*extent)[0];
    const double ymin = (*extent)[1];
    const double xmax = (*extent)[2];
    const double ymax = (*extent)[3];
    if (!(xmin <= xmax && ymin <= ymax)) {
        return std::nullopt;
    }
    const std::optional<Domain> domain = crs_coordinate_domain(crs);
    if (!domain.has_value()) {
        return std::nullopt;
    }
    const double dminx = (*domain)[0];
    const double dminy = (*domain)[1];
    const double dmaxx = (*domain)[2];
    const double dmaxy = (*domain)[3];
    const double eps =
        kDomainEpsilon *
        std::max({1.0, std::fabs(dmaxx - dminx), std::fabs(dmaxy - dminy)});
    const bool fits = xmin >= dminx - eps && xmax <= dmaxx + eps &&
                      ymin >= dminy - eps && ymax <= dmaxy + eps;
    if (fits) {
        return std::nullopt;
    }
    DomainMismatch out;
    const std::string normalized = normalize_crs(crs);
    out.crs = !normalized.empty() ? normalized : crs.value_or("");
    out.extent = *extent;
    out.domain = *domain;
    return out;
}

CRSInference infer_crs_from_extent(const std::optional<Domain>& extent) {
    if (!extent.has_value() ||
        !((*extent)[0] <= (*extent)[2] && (*extent)[1] <= (*extent)[3])) {
        return {"", "数据坐标范围不可用——不推断"};
    }
    if (!coordinate_domain_mismatch(std::string("EPSG:4326"), extent)
             .has_value()) {
        return {"EPSG:4326", "数据坐标完全落在经纬度域内（地理坐标）"};
    }
    return {"", "数据坐标超出经纬度域（本地/投影坐标）——保持本地"};
}

}  // namespace pwb::mapping
